tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
 use std::env;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
 
 use colored::*;
-use compiletest_rs as compiletest;
+use crossbeam::queue::SegQueue;
+use regex::Regex;
 
 fn miri_path() -> PathBuf {
     PathBuf::from(option_env!("MIRI").unwrap_or(env!("CARGO_BIN_EXE_miri")))
 }
 
-fn run_tests(mode: &str, path: &str, target: &str) {
+fn run_tests(mode: Mode, path: &str, target: &str) {
     let in_rustc_test_suite = option_env!("RUSTC_STAGE").is_some();
+
+    if in_rustc_test_suite {
+        std::env::set_var("MIRI_SKIP_UI_CHECKS", "1");
+    }
+
     // Add some flags we always want.
     let mut flags = Vec::new();
-    flags.push("--edition 2018".to_owned());
+    flags.push("--edition".to_owned());
+    flags.push("2018".to_owned());
     if in_rustc_test_suite {
         // Less aggressive warnings to make the rustc toolstate management less painful.
         // (We often get warnings when e.g. a feature gets stabilized or some lint gets added/improved.)
         flags.push("-Astable-features".to_owned());
     } else {
-        flags.push("-Dwarnings -Dunused".to_owned()); // overwrite the -Aunused in compiletest-rs
+        flags.push("-Dwarnings".to_owned());
+        flags.push("-Dunused".to_owned()); // overwrite the -Aunused in compiletest-rs
     }
     if let Ok(sysroot) = env::var("MIRI_SYSROOT") {
-        flags.push(format!("--sysroot {}", sysroot));
+        flags.push("--sysroot".to_string());
+        flags.push(sysroot);
     }
     if let Ok(extra_flags) = env::var("MIRIFLAGS") {
-        flags.push(extra_flags);
+        for flag in extra_flags.split_whitespace() {
+            flags.push(flag.to_string());
+        }
     }
+    flags.push("-Zui-testing".to_string());
+    flags.push("--target".to_string());
+    flags.push(target.to_string());
+    let target = target.to_string();
 
-    let flags = flags.join(" ");
-    eprintln!("   Compiler flags: {}", flags);
+    eprintln!("   Compiler flags: {:?}", flags);
 
-    // The rest of the configuration.
-    let mut config = compiletest::Config::default().tempdir();
-    config.mode = mode.parse().expect("Invalid mode");
-    config.rustc_path = miri_path();
-    if let Some(lib_path) = option_env!("RUSTC_LIB_PATH") {
-        config.run_lib_path = PathBuf::from(lib_path);
-        config.compile_lib_path = PathBuf::from(lib_path);
-    }
-    config.filters = env::args().nth(1).into_iter().collect();
-    config.host = get_host();
-    config.src_base = PathBuf::from(path);
-    config.target = target.to_owned();
-    config.target_rustcflags = Some(flags);
-    compiletest::run_tests(&config);
-}
+    let grab_entries =
```

Comment:
There's a lot of new code here without many comments. Please add some comments explaining what is going on.